UI/window-basic-preview-controls.cpp

```diff
@@ -0,0 +1,158 @@
+{
+	previewScale = scale;
+	UpdateScaleLabel();
+}
+
+void OBSPreviewScalingLabel::UpdateScaleLabel()
+{
+	float previewScalePercent = floor(100.0f * previewScale);
+	setText(QString::number(previewScalePercent) + "%");
+}
+
+/* Preview Scaling ComboBox */
+OBSPreviewScalingComboBox::OBSPreviewScalingComboBox(QWidget *parent)
+	: QComboBox(parent)
+{
+}
+
+void OBSPreviewScalingComboBox::on_previewFixedScalingChanged(bool isFixed)
+{
+	if (isFixedScaling() == isFixed)
+		return;
+
+	setFixedScaling(isFixed);
+	UpdateSelection();
+}
+
+void OBSPreviewScalingComboBox::on_canvasResized(uint32_t width,
+						 uint32_t height)
+{
+	setCanvasSize(width, height);
+	UpdateCanvasText();
+}
+
+void OBSPreviewScalingComboBox::on_outputResized(uint32_t width,
+						 uint32_t height)
+{
+	setOutputSize(width, height);
+
+	bool canvasMatchesOutput = output_width == canvas_width &&
+				   output_height == canvas_height;
+
+	setScaleOutputEnabled(!canvasMatchesOutput);
+	UpdateOutputText();
+}
+
+void OBSPreviewScalingComboBox::on_previewScaleChanged(float scale)
+{
+	previewScale = scale;
+
+	if (isFixedScaling()) {
+		UpdateSelection();
+		UpdateAllText();
+	} else {
+		UpdateScaledText();
+	}
+}
+
+void OBSPreviewScalingComboBox::setScaleOutputEnabled(bool show)
+{
+	if (scaleOutputEnabled == show)
+		return;
+
+	scaleOutputEnabled = show;
+
+	if (scaleOutputEnabled) {
+		addItem(QTStr("Basic.MainMenu.Edit.Scale.Output"));
+	} else {
+		removeItem(2);
+	}
+}
+
+void OBSPreviewScalingComboBox::UpdateAllText()
+{
+	UpdateCanvasText();
+	UpdateOutputText();
+	UpdateScaledText();
+}
+
+void OBSPreviewScalingComboBox::UpdateCanvasText()
+{
+	QString text = QTStr("Basic.MainMenu.Edit.Scale.Canvas");
+	text = text.arg(QString::number(canvas_width),
+			QString::number(canvas_height));
+	setItemText(1, text);
+}
+
+void OBSPreviewScalingComboBox::UpdateOutputText()
+{
+	if (scaleOutputEnabled) {
+		QString text = QTStr("Basic.MainMenu.Edit.Scale.Output");
+		text = text.arg(QString::number(output_width),
+				QString::number(output_height));
+		setItemText(2, text);
+	}
+}
+
+void OBSPreviewScalingComboBox::UpdateScaledText()
+{
+	QString text = QTStr("Basic.MainMenu.Edit.Scale.Manual");
+	text = text.arg(QString::number(floor(canvas_width * previewScale)),
+			QString::number(floor(canvas_height * previewScale)));
+	setPlaceholderText(text);
+}
+
+void OBSPreviewScalingComboBox::UpdateSelection()
+{
+	float outputScale = float(output_width) / float(canvas_width);
+
+	if (!isFixedScaling()) {
+		setCurrentIndex(0);
+	} else {
+		if (previewScale == 1.0f) {
+			setCurrentIndex(1);
+		} else if (scaleOutputEnabled &&
+			   (previewScale == outputScale)) {
+			setCurrentIndex(2);
+		} else {
+			setCurrentIndex(-1);
+		}
+	}
+}
+
+void OBSPreviewScalingComboBox::setCurrentIndex(int index)
+{
+	if (currentIndex() != index) {
+		bool blocked = blockSignals(true);
+		QComboBox::setCurrentIndex(index);
+		blockSignals(blocked);
+	}
+}
```

Comment:
Without even going into the potential dangers of overloading non-virtual methods (`setCurrentIndex` in this case), I think this can be solved much more easily by just doing something like `QSignalBlocker blocker(this);` at the beginning of `UpdateSelection` and removing the `setCurrentIndex` overload completely.

Reply:
Fixed